### PR TITLE
Fix StoreOpConversion bug and add test_where to test_core_amd.py.

### PR DESCRIPTION
**Pull request by rsanthanam-amd**
Will need to revisit these changes once we implement vectorized loads and stores.

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 1]
```diff
@@ -1229,7 +1229,7 @@ struct StoreOpConversion
         asmArgs.emplace_back(llWord, constraint);
       }
 #ifdef USE_ROCM
-      store(asmArgs[vecStart].first, ptrElems[vecStart]);
```
Comment (rsanthanam-amd) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1232]: This is the fix.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 2]
```diff
@@ -1007,10 +1007,7 @@ struct LoadOpConversion
 
 #ifdef USE_ROCM
       SmallVector<Type> retTys(nWords, IntegerType::get(getContext(), width));
-      Type retTy = retTys.size() > 1
```
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1010]: @rsanthanam-amd There is no description in the PR. Could you explain why the change works? The original logic seems to stem from NV path which uses inline asm. Since we are not using inline asm so perhaps it caused some differences and I'd like to understand why.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 3]
```diff
@@ -1229,7 +1226,7 @@ struct StoreOpConversion
         asmArgs.emplace_back(llWord, constraint);
       }
 #ifdef USE_ROCM
-      store(asmArgs[vecStart].first, ptrElems[vecStart]);
```
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1232]: @rsanthanam-amd Again I want an explanation on why this change fixes stuffs. Looking at the logic essentially we are ditching the for loop of `vecStart`.

**Review by whchung — CHANGES_REQUESTED** [review]
Justification of changes proposed is needed.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 3 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1232]: There is a nested for-loop which populates the asmArgs data structure.  

This data structure is recreated for each vecStart loop iteration and for ROCm, the nWords loop condition is always 1 since we are only emitting a single store instruction per vecStart loop iteration.

My previous code incorrectly indexed into the asmArgs data structure with vecStart loop induction variable.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 2 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1010]: For ROCm, we only emit a single load.

With the previous code, the type was specified as an aggregate if the size was greater than 1.  This was causing LLVM to give an error because it was expecting a singleton type but got an aggregate type.

This changes forces the type to always be a singleton.

Comment (whchung) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1010]: Several follow-up questions:
- Can we be sure that 1 single load would suffice?
- What would be possible values of `nWords` for different data types?
- What would be possible values of `width` for different data types?
- When `nWords` is larger than 1, the original logic creates a `LLVMStructType`, set it as `retTy`, and later on use `extract_val` to extract elements from it. New logic introduced here essentially skip those logic in "extract and store return values" section in the later part of the function.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 3 above)
Comment (whchung) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1232]: Similar to my questions on load above:
- Can we be sure that 1 single load would suffice?
- What would be possible values of `nWords` for different data types?
- If we are sure `nWords` would always be 1, shall the loop which creates `asmArgs` be changed? Its original logic could be NV-specific where PTX inline asm may require the type of each argument be specified.
- In the earlier logic we change bitcast from `valArgTy` to `valueElemTy` between NV and AMD path, what are their values(types)? I guess `valueElemTy` could be a vector type, while `valArgTy` is always a scalar type? Is this understanding correct?
- Ultimately I feel we probably shouldn't depend on `asmArgs` vector anymore, as it no longer holds data we really care.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 2 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1010]: I can't be sure if 1 single load would suffice but since we are only doing scalar (not vectorized) loads, it would seem to be the case.

For scalar loads, nWords should always be 1.

Width is equivalent to the number of bits in the data type so it would be 8 for uint8 and 32 for float32 for example.

The dependence on LLVMStructType is not good as you suggest.  I will submit a change which removes ROCm's dependence on this.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 3 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1232]: I can't be sure that a single store would suffice but since we are not doing vectorized stores, it seems so.

nWords should always be 1 for scalar store.

"I guess valueElemTy could be a vector type, while valArgTy is always a scalar type" - I believe this is correct.

I will remove the use of the asmArgs vector for ROCm.

**Comment by whchung** [comment]
@rsanthanam-amd After you come up with a revised implementation, please update the description of the PR. I think we need a TODO item to record that vectorization may not be enabled right now, and shall be left as a subsequent PR.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 2 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1010]: pushed some changes which eliminate our dependence on asmArgs

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 3 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1232]: pushed some changes which eliminate our dependence on asmArgs

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp [hunk 4]
```diff
@@ -1219,17 +1217,24 @@ struct StoreOpConversion
...
         std::string constraint =
             (width == 64) ? "l" : ((width == 32) ? "r" : "c");
         asmArgs.emplace_back(llWord, constraint);
       }
 #ifdef USE_ROCM
-      store(asmArgs[vecStart].first, ptrElems[vecStart]);
+      // TODO: need to revisit this when implementing vectorized stores
+      Value llWord = rewriter.create<LLVM::UndefOp>(loc, wordTy);
+      Value elem = valueElems[vecStart];
+      if (elem.getType().isInteger(1))
```
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1229]: `elem.getType().getBitWidth() < 8`?

**Review by whchung — COMMENTED** [review]
one more minor change and I think the PR is good to go.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp (hunk 4 above)
Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1229]: i'm not sure i understand the concern.

this transformation handles 1 bit integer data types (bits) i believe and the next larger data width is 8 bits which does not need special handling.

Comment (whchung) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1229]: If we are certain `i1` is the only plausible type then I'm fine with existing logic. Just not sure if we'll see `i2~7`.

Comment (rsanthanam-amd) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp]: fixed.

you were right - i2-7 are not built in types but users can define custom integer types with those bit widths.

**Review by whchung — APPROVED** [review]
